assemble/Particle_Diagnostics.F90

```diff
@@ -0,0 +1,1420 @@
+          call get_option(trim(complete_field_path(s_field%option_path))// "/algorithm/particle_group/particle_attribute/name", lattribute)
+          att_n=0
+          ewrite(2,*) "Calculate diagnostic field from particle group: ", trim(lgroup), ", attribute: ", trim(lattribute)
+       else if (have_option(trim(complete_field_path(s_field%option_path))// "/algorithm/particle_group/particle_attribute_array")) then
+          call get_option(trim(complete_field_path(s_field%option_path))// "/algorithm/particle_group/particle_attribute_array/name", lattribute)
+          call get_option(trim(complete_field_path(s_field%option_path))// "/algorithm/particle_group/particle_attribute_array/integer_val", att_n)
+          ewrite(2,*) "Calculate diagnostic field from particle group: ", trim(lgroup), ", attribute_array: ", trim(lattribute), ", integer_val: ", att_n
+       end if
+    else
+       ewrite(2,*) "Calculate diagnostic field from number of particles on particle group: ", trim(lgroup)
+    end if
+
+    !Initialize field as 0
+    s_field%val(:) = 0
+
+    xfield=>extract_vector_field(states(1), "Coordinate")
+
+    nodes => ele_nodes(s_field, 1)
+    if (xfield%dim+1/=size(nodes)) then
+       FLAbort("Can only generate particle diagnostic fields for a P1CV mesh")
+    end if
+
+    ! contribution of local particles to non-owned nodes are summed
+    ! into the owner in the halo_accumulate calls below
+    ! we might be safe to assume we only need to add into halo 1 nodes (as these
+    ! are the only ones that make up owned elements), but let's include halo 2 to be sure
+    ! Only run this if nprocs > 1
+    nprocs = getnprocs()
+    if (nprocs>1) then
+       halo => s_field%mesh%halos(2)
+    end if
+
+    if (.not. allocated(particle_lists)) then
+       !Particles not yet setup, Initial field value will be 0
+       return
+    else
+       !Particles are initialized, call subroutine to get relevant particle arrays and attributes
+       if (is_ratio) then
+          call get_particle_arrays(lgroup, group_arrays, group_attribute, att_n=att_n, lattribute=lattribute)
+       else
+          call get_particle_arrays(lgroup, group_arrays)
+       end if
+       
+
+       !Allocate arrays to store summed attribute values and particle counts at nodes
+       if (is_ratio) then
+          allocate(node_values(node_count(s_field)))
+          node_values(:) = 0
+       end if
+       allocate(node_part_count(node_count(s_field)))
+       node_part_count(:) = 0
+
+       if (is_ratio .and. have_option(trim(complete_field_path(s_field%option_path))// "/algorithm/interpolation/weighted_distance")) then
+
+          !Loop over particle arrays
+          do i = 1,size(group_arrays)
+             particle => particle_lists(group_arrays(i))%first
+             if (associated(particle)) then !Only work on arrays if local_particles exist on this processor
```

Comment:
You could replace this with `if (.not. associated(particle)) cycle`
Then on the next line you always allocate `local_crds` (if it gets there), and you don't have to check `if (allocated(local_crds))` after the particle loop.